core/src/main/java/com/google/cloud/sql/core/SqlAdminApiService.java

```diff
@@ -0,0 +1,230 @@
+/*
+ * Copyright 2023 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.sql.core;
+
+import com.google.api.client.googleapis.json.GoogleJsonResponseException;
+import com.google.api.client.http.HttpRequestInitializer;
+import com.google.api.services.sqladmin.SQLAdmin;
+import com.google.api.services.sqladmin.model.ConnectSettings;
+import com.google.api.services.sqladmin.model.GenerateEphemeralCertRequest;
+import com.google.api.services.sqladmin.model.GenerateEphemeralCertResponse;
+import com.google.api.services.sqladmin.model.IpMapping;
+import com.google.auth.oauth2.GoogleCredentials;
+import com.google.auth.oauth2.OAuth2Credentials;
+import com.google.cloud.sql.ApiClientFactory;
+import com.google.common.base.CharMatcher;
+import com.google.common.io.BaseEncoding;
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.security.KeyPair;
+import java.security.cert.Certificate;
+import java.security.cert.CertificateException;
+import java.security.cert.CertificateFactory;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+
+public class SqlAdminApiService {
+
+  private static final String SQL_LOGIN_SCOPE = "https://www.googleapis.com/auth/sqlservice.login";
+  SQLAdmin apiClient;
+
+  public SqlAdminApiService(ApiClientFactory sqlAdminApiFactory,
+      HttpRequestInitializer credential) {
+    apiClient = (SQLAdmin) sqlAdminApiFactory.create(credential);
+  }
+
+  static void checkDatabaseCompatibility(ConnectSettings instanceMetadata, boolean iamAuth,
```

Comment:
Let's avoid static functions generally as we'll always be dealing with instances in our code.